internal/caveats/debug.go

```diff
@@ -0,0 +1,149 @@
+package caveats
+
+import (
+	"fmt"
+	"maps"
+	"strconv"
+	"strings"
+
+	"google.golang.org/protobuf/types/known/structpb"
+
+	"github.com/authzed/spicedb/pkg/caveats"
+	"github.com/authzed/spicedb/pkg/genutil/mapz"
+	corev1 "github.com/authzed/spicedb/pkg/proto/core/v1"
+	"github.com/authzed/spicedb/pkg/spiceerrors"
+)
+
+// BuildDebugInformation returns a human-readable string representation of the given
+// ExpressionResult and a Struct representation of the context values used in the expression.
+func BuildDebugInformation(exprResult ExpressionResult) (string, *structpb.Struct, error) {
+	// If a concrete result, return its information directly.
+	if concrete, ok := exprResult.(*caveats.CaveatResult); ok {
+		exprString, err := concrete.ParentCaveat().ExprString()
+		if err != nil {
+			return "", nil, err
+		}
+
+		contextStruct, err := concrete.ContextStruct()
+		if err != nil {
+			return "", nil, err
+		}
+
+		return exprString, contextStruct, nil
+	}
+
+	// Collect parameters which are shared across expressions.
+	syntheticResult := exprResult.(syntheticResult)
```

Comment:
just in case we ever create a third type of `ExpressionResult`, let's check the type assertion actually succeeded